03-attribute-operations.Rmd

```diff
@@ -499,14 +507,211 @@ world_data = world %>% st_set_geometry(NULL)
+A focal operation applies an aggregation function to all cells within the specified neighborhood, uses the corresponding output as the new value for the the central cell, and moves on to the next central cell (Figure @ref(fig:focal_example)).
+Other names for this operation are spatial filtering and convolution [@burrough_principles_2015].
+
+In R, we can use the `focal()` function to perform spatial filtering. 
+We define the shape of the moving window with a `matrix` whose values correspond to weights.
+Secondly, the the `fun` argument lets us specify the function we wish to apply to this neighborhood.
+Here, we choose the minimum, but of course we can use any other function such as the the sum, the mean, the median, the mode, the maximum or the variance.
+
+```{r, eval = FALSE}
+r_focal = focal(r, w = matrix(1, nrow = 3, ncol = 3), fun = min)
+```
+
+```{r focal_example, echo = FALSE, fig.cap = "Input raster (left) and resulting output raster (right) due to a focal operation - summing up 3-by-3 windows"}
+knitr::include_graphics("figures/03_focal_example.png")
+```
+
+We can quickly check if the output meets our expectations.
+In our example, the minimum value has to be always the upper left corner of the moving window (remember we have created the input raster by rowwise incrementing the cell values by one starting at the upper left corner).
+Of course, the `focal()`-function has computed the correct result.
+In this example, our weighting matrix consists only of 1s.
+This means each cell has the same weight on the output.
+You can change this by specifying different weights.
+
+Focal functions or filters play a dominant role in image processing.
+Low-pass or smoothing filters use the mean function to remove extremes.
+In the case of categorical data, we can replace the mean with the mode, which is the most common value.
+By contrast, high-pass filters accentuate features.
+The line detection Laplace and Sobel filters might serve as an example here.
+Check the `focal()` help page how to use them in R.
+
+Also, terrain processing uses heavily focal functions.
+Think, for instance, of the calculation of the slope, aspect and flow directions.
+The `terrain()` function lets you compute a few of these terrain characteristics but has not implemented all popular methods
+For example, the Zevenbergen and Thorne method to compute the slope is missing.
+Equally, many other terrain and GIS functions are **not** implemented in R such as curvatures, contributing areas, wetness indexes, and many more.
+Fortunately, desktop GIS commonly provide these algorithms.
+In Chapter 13 we will learn how to access GIS functionality from within R.
+<!-- Reference 13-gis chapter -->
+
+**Zonal and global operations**
```

Comment:
Not finished the rest yet, still have to write this. Will do so tomorrow.

Reply:
Great piece of continuous technical prose.